src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Migrations/1.sql

```diff
@@ -1,43 +1,332 @@
+--         * If false, an error is thrown if the resource does not already exist
+--     @isDeleted
+--         * Whether this resource marks the resource as deleted
+--     @updatedDateTime
+--         * The last modifed time in the the resource
+--     @keepHistory
+--         * Whether the existing version of the resource should be preserved
+--     @requestMethod
+--         * The HTTP method/verb used for the request
+--     @rawResource
+--         * A compressed UTF16-encoded JSON document  
+--
+-- RETURN VALUE
+--         The version of the resource as a result set. Will be empty if no insetion was done.
+--
+CREATE PROCEDURE dbo.UpsertResource
+    @resourceTypeId smallint,
+	@resourceId varchar(64),
+    @eTag int = NULL,
+    @allowCreate bit,
+    @isDeleted bit,
+    @updatedDateTime datetimeoffset(7),
+    @keepHistory bit,
+    @requestMethod varchar(10),
+	@rawResource varbinary(max)
+AS
+	SET XACT_ABORT ON
+	BEGIN TRANSACTION
+
+    DECLARE @previousVersion TABLE(  
+        ResourceSurrogateId bigint NOT NULL,  
+        Version int NOT NULL);  
+
+    if (@keepHistory = 1) BEGIN
+        -- Preserve the existing version, marking it as history
+        UPDATE dbo.Resource WITH (UPDLOCK, HOLDLOCK)
+        SET IsHistory = 1
+        OUTPUT inserted.ResourceSurrogateId,  
+                inserted.Version
+        INTO @previousVersion
+        WHERE ResourceTypeId = @resourceTypeId AND ResourceId = @resourceId AND IsHistory = 0 AND (@isDeleted = 0 OR IsDeleted = 0)
+    END
+    ELSE BEGIN
+        -- Delete the previous version
+        DELETE FROM dbo.Resource WITH (UPDLOCK, HOLDLOCK)
+        OUTPUT deleted.ResourceSurrogateId,  
+                deleted.Version
+        INTO @previousVersion
+        WHERE ResourceTypeId = @resourceTypeId AND ResourceId = @resourceId AND IsHistory = 0 AND (@isDeleted = 0 OR IsDeleted = 0)
+    END
+
+    DECLARE @version int;
+
+    if (@@ROWCOUNT = 0) BEGIN
+        IF (@isDeleted = 1) BEGIN
+            -- Either a previous version does not exist or it is already an "IsDeleted" version
+            COMMIT TRANSACTION
+            RETURN
+        END
+
+        IF (@allowCreate = 0) BEGIN
+            THROW 50404, 'Resource does not exist and create is not allowed', 1;
+        END
+
+        SET @version = 1
+    END
+    ELSE BEGIN
+        -- There is a previous version
+        SET @version = (select (Version + 1) from @previousVersion)
+    END
+
+
+    IF (@etag IS NOT NULL AND @etag <> (@version - 1)) BEGIN
+        THROW 50412, 'Precondition failed', 1;
+    END
+
+    DECLARE @resourceSurrogateId bigint = NEXT VALUE FOR dbo.ResourceSurrograteIdSequence
+
+    INSERT INTO dbo.Resource 
+        (ResourceTypeId, ResourceId, Version, IsHistory, ResourceSurrogateId, LastUpdated, IsDeleted, RequestMethod, RawResource)
+    VALUES 
+        (@resourceTypeId, @resourceId, @version, 0, @resourceSurrogateId, CONVERT(datetime2(7), @updatedDateTime), @isDeleted, @requestMethod, @rawResource)
```

Comment:
This `LastUpdated` field will be different than what is in the resource. Is that acceptable/expected?

Reply:
It is actually the same. We pass this in as a parameter, and it is set on the resource in BaseResourceHandler.CreateResourceWrapper

Reply:
I missed that.